README
DBIx-NinjaORM

Flexible Perl ORM for easy transitions from inline SQL to objects.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc DBIx::NinjaORM

You can also look for information at:

    GitHub's request tracker (report bugs here)
        https://github.com/guillaumeaubert/DBIx-NinjaORM/issues

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/DBIx-NinjaORM

    CPAN Ratings
        http://cpanratings.perl.org/d/DBIx-NinjaORM

    Search CPAN
        https://metacpan.org/release/DBIx-NinjaORM


LICENSE AND COPYRIGHT

Copyright (C) 2009-2013 Guillaume Aubert

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/



Changes
Revision history for DBIx-NinjaORM

v3.0.2  2014-04-20
        - Fixed "package version matches dist version" for CPANTS.
        - Fixed license declared in meta files.
        - Added option to customize the dumper function used (GH-3).
        - General cleanup.

v3.0.1  2013-08-25
        - Fixed double-planning tests when memcache isn't running (GH-21).
        - Reordered new tests.

v3.0.0  2013-08-25
        - SEE UPGRADE NOTES: refactored class configuration into
          DBIx::NinjaORM::StaticClassInfo (GH-15).
        - SEE UPGRADE NOTES: added get_info() and deprecated some accessors
          which will be removed in a next version (GH-19).
        - Added document to gather upgrade instructions between versions.
        - Added ability to use real date types for 'created' and 'modified'
          fields (GH-2).
        - Set up mechanism to retrieve and cache schema information, to power
          new features of NinjaORM that rely on knowing the table structures (GH-16).
        - Added the ability to select explicitely or exclude fields from the
          underlying table as part of retrieve_list(), as opposed to table.* (GH-14).
        - Fixed bug where an object ID would exist in the list cache but have
          been deleted from the database and return undef instead of being
          skipped in the results of retrieve_list().
        - Fixed bug when retrieving cache misses by ID, now passing the correct
          arguments for dbh/show_queries.
        - Refactored retrieve_list() argument handling for list cache key
          generation.
        - Code cleanup.
        - Better test coverage (including GH-10, GH-18).
				* Massive tests cleanup, reusable test classes are now in
          t/lib/TestSubclass/.

v2.5.1  2013-07-11
        - Fixed retrieve_list() when a 'created' field is not present (GH-17).
          Thanks Kate Kirby!
        - Added tests for retrieve_list() and update() without created/modified
          fields.
        - Documented pagination for retrieve_list().
        - Added release tests for file encoding.

v2.5.0  2013-05-22
        - Added 'like' and 'not_like' operators to retrieve_list() (GH-13).
          Thanks Brian Voorhes!

v2.4.4  2013-03-30
        - Replaced Test::NoWarnings with Test::FailWarnings.
        - Added new release tests.
        - Moved release tests to xt/, updated Build.PL.

v2.4.3  2013-03-21
        - (GitHub #7) Fixed warnings in Perl 5.16.
        - (GitHub #8) Fixed manifest.
        - (GitHub #9) Fixed prerequisite.

v2.4.2  2013-03-18
        - Moved request tracker from RT to GitHub.
        - Changed links from Search CPAN to MetaCPAN.

v2.4.1  2013-02-24
        - (RT #83513) Fixed incorrect croak() in verbose mode.
        - (RT #83514) Don't delete cache if no cache key can be determined.
        - Modified validate_data() to check for existence, not just defineness
          of keys, as undefined keys would still impact insert/update.

v2.4.0  2013-02-19
        - Fixed detecting invalid filtering criteria [].

v2.3.2  2013-01-27
        - Fixed propagation of DBI failures in update().
        - Added tests to cover DBI failures for insert(), update(), and
          remove().
        - Added 'allow_subclassing' option for retrieve_list().
        - Added new tests for pagination in retrieve_list_nocache().
        - Cleaned up tests and code.

v2.3.1  2013-01-05
        - Fixed call to selectall_arrayref() in retrieve_list_nocache() and
          added error handling (thanks Jacob Rose).
        - Added tests to cover locking with joins in retrieve_list_nocache().
        - Fixed list of locked IDs in retrieve_list_nocache().
        - retrieve_list_nocache() now ignores lock=1 for SQLite.

v2.3.0  2012-12-04
        - Refactored to use Log::Any instead of carp/confess.

v2.2.1  2012-12-04
        - Split limit and offset, for PostgreSQL.

v2.2.0  2012-12-04
        - Modified assert_dbh() to allow coderefs that return a DBI::db object
          in either default_dbh from static_class_information() or from its
          first parameter.
        - Added tests for the pagination feature of retrieve_list().
        - Added tests for assert_dbh().
        - Cleaned up warning that should only show up in verbose mode.

v2.1.3  2012-12-03
        - Fixed pagination in retrieve_list().

v2.1.2  2012-11-30
        - Added tests for remove() and commit().
        - If custom clauses are present, filters are not required anymore.
        - Checked filters argument in new() to detect issues early.
        - Sorted filters in parse_filtering_criteria() to make output easily
          testable.

v2.1.1  2012-11-26
        - Ignored Devel::Cover output files.
        - Added Perl 5.10 as a requirement.

v2.1.0  2012-11-22
        - Refactored retrieve_list() and new() to use a hashref as first
          argument, like insert() and update().
        - Reorganized methods alphabetically.

v2.0.0  2012-11-19
        - Generalized code, added more tests and documentation.
        - First public release.

v1.9.1  2012-10-24
        - Added 'not_null' operator.

v1.9.0  2012-10-18
        - Updated to use two separate cache times, one for the translation from
          arguments to IDs (list cache) and one for the objects (object cache).
          Thanks Jennifer Pinkham!

v1.8.4  2012-10-02
        - Added get_object_cache_keys().

v1.8.3  2012-10-01
        - Added "set" parameter to update() to allow specifying complex set
          clauses.

v1.8.2  2012-09-07
        - Added option to trace cache operations.

v1.8.1  2012-09-06
        - Adjusted the order of when we invalidate the object cache in update().
        - Forced the cache keys and values to be lower case, to account for
          case-insensitive collations.

v1.8.0  2012-08-30
        - Implemented cache of the translation from arguments to IDs in
          retrieve_list(). Thanks Kate Kirby for pair programming this change!

v1.7.7  2012-05-22

t/LocalTest.pm
package LocalTest;

use strict;
use warnings;

use DBI;
use Test::More;


=head1 NAME

LocalTest - Test functions for L<DBIx::NinjaORM>.


=head1 VERSION

Version 1.0.0

=cut

our $VERSION = '1.0.0';


=head1 SYNOPSIS

	use lib 't/';
	use LocalTest;
	
	my $dbh = LocalTest::ok_database_handle();


=head1 FUNCTIONS

=head2 get_database_handle()

Create a database handle.

	my $dbh = LocalTest::get_database_handle();

=cut

sub get_database_handle
{
	$ENV{'NINJAORM_DATABASE'} ||= 'dbi:SQLite:dbname=t/test_database||';
	
	my ( $database_dsn, $database_user, $database_password ) = split( /\|/, $ENV{'NINJAORM_DATABASE'} );
	
	my $database_handle = DBI->connect(
		$database_dsn,
		$database_user,
		$database_password,
		{
			RaiseError => 1,
		}
	);
	
	# If it's SQLite, we need to turn on foreign keys support.
	if ( defined( $database_handle ) )
	{
		my $database_type = get_database_type( $database_handle );
		if ( defined( $database_type ) && ( $database_type eq 'SQLite' ) )
		{
			$database_handle->do( 'PRAGMA foreign_keys = ON' );
		}
	}
	
	return $database_handle
}


=head2 ok_database_handle()

Verify that a database handle can be created, and return it.

	my $dbh = LocalTest::ok_database_handle();

=cut

sub ok_database_handle
{
	ok(
		defined(
			my $database_handle = get_database_handle()
		),
		'Create connection to a database.',
	);
	
	my $database_type = $database_handle->{'Driver'}->{'Name'} || '';
	note( "Testing $database_type database." );
	
	return $database_handle;
}


=head2 get_database_type()

Return the name of the driver used by the database handle.

	my $database_type = LocalTest::get_database_type( $database_handle );

=cut

sub get_database_type
{
	my ( $dbh ) = @_;
	
	return $dbh->{'Driver'}->{'Name'};
}


=head2 ok_database_type()

Verify that the database type is supported, and return it.

	my $database_type = LocalTest::ok_database_type( $database_handle );

=cut

sub ok_database_type
{
	my ( $dbh ) = @_;
	
	my $type = get_database_type( $dbh ) || '';
	
	like(
		$type,
		qr/^(?:mysql|SQLite)$/,
		"Database type '$type' is supported.",
	);
	
	return $type;
}


=head2 get_memcache()

Get a memcache object.

	my $memcache = LocalTest::get_memcache();

=cut

sub get_memcache
{
	return
		try
		{
			eval 'use Cache::Memcached::Fast';
			return Cache::Memcached::Fast->new(
				{
					servers =>
					[
						'localhost:11211',
					],
				}
			);
		}
		catch
		{
			return $_;
		};
}


=head2 ok_memcache()

Verify that memcache is running, and skip all tests if it isn't.

	my $memcache = LocalTest::ok_memcache();

=cut

sub ok_memcache
{
	# Verify if Cache::Memcached::Fast is installed.
	eval 'use Cache::Memcached::Fast';
	plan( skip_all => 'Cache::Memcached::Fast required to test rate limiting.' )
		if $@;
	
	my $memcache = LocalTest::get_memcache();
	
	# Verify that memcache is configured and running.
	plan( skip_all => 'Memcache is not running or configured on this machine, cannot test rate limiting' )
		if !defined( $memcache) || !$memcache->set( 'test_ninja_orm', 1, time() + 10 );
	
	return $memcache;
}


=head1 AUTHOR

Guillaume Aubert, C<< <aubertg at cpan.org> >>.


=head1 BUGS

Please report any bugs or feature requests to C<bug-dbix-ninjaorm at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=DBIx-NinjaORM>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc LocalTest


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DBIx-NinjaORM>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-NinjaORM>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-NinjaORM>

=item * Search CPAN

L<http://search.cpan.org/dist/DBIx-NinjaORM/>

=back


=head1 COPYRIGHT & LICENSE

Copyright 2009-2012 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


t/lib/LocalTest.pm
package LocalTest;

use strict;
use warnings;

use Carp;
use Data::Validate::Type;
use DBI;
use Test::More;
use Try::Tiny;


=head1 NAME

LocalTest - Test functions for L<DBIx::NinjaORM>.


=head1 VERSION

Version 3.0.2

=cut

our $VERSION = '3.0.2';


=head1 SYNOPSIS

	use lib 't/lib';
	use LocalTest;

	my $dbh = LocalTest::ok_database_handle();


=head1 FUNCTIONS

=head2 get_database_handle()

Create a database handle.

	my $dbh = LocalTest::get_database_handle();

=cut

sub get_database_handle
{
	$ENV{'NINJAORM_DATABASE'} ||= 'dbi:SQLite:dbname=t/test_database||';

	my ( $database_dsn, $database_user, $database_password ) = split( /\|/, $ENV{'NINJAORM_DATABASE'} );

	my $database_handle = DBI->connect(
		$database_dsn,
		$database_user,
		$database_password,
		{
			RaiseError => 1,
		}
	);

	# If it's SQLite, we need to turn on foreign keys support.
	if ( defined( $database_handle ) )
	{
		my $database_type = get_database_type( $database_handle );
		if ( defined( $database_type ) && ( $database_type eq 'SQLite' ) )
		{
			$database_handle->do( 'PRAGMA foreign_keys = ON' );
		}
	}

	return $database_handle
}


=head2 ok_database_handle()

Verify that a database handle can be created, and return it.

	my $dbh = LocalTest::ok_database_handle();

=cut

sub ok_database_handle
{
	ok(
		defined(
			my $database_handle = get_database_handle()
		),
		'Create connection to a database.',
	);

	my $database_type = $database_handle->{'Driver'}->{'Name'} || '';
	note( "Testing $database_type database." );

	return $database_handle;
}


=head2 get_database_type()

Return the name of the driver used by the database handle.

	my $database_type = LocalTest::get_database_type( $database_handle );

=cut

sub get_database_type
{
	my ( $dbh ) = @_;

	croak 'The first argument of get_database_type() must be a database handle'
		if ! Data::Validate::Type::is_instance( $dbh, class => 'DBI::db' );

	return $dbh->{'Driver'}->{'Name'};
}


=head2 ok_database_type()

Verify that the database type is supported, and return it.

	my $database_type = LocalTest::ok_database_type( $database_handle );

=cut

sub ok_database_type
{
	my ( $dbh ) = @_;

	my $type = get_database_type( $dbh ) || '';

	like(
		$type,
		qr/^(?:mysql|SQLite|Pg)$/,
		"Database type '$type' is supported.",
	);

	return $type;
}


=head2 get_memcache()

Get a memcache object.

	my $memcache = LocalTest::get_memcache();

=cut

sub get_memcache
{
	return
		try
		{
			eval 'use Cache::Memcached::Fast';
			die 'Cache::Memcached::Fast is not installed on this system'
				if $@;

			return Cache::Memcached::Fast->new(
				{
					servers =>
					[
						'localhost:11211',
					],
				}
			);
		}
		catch
		{
			return undef;
		};
}


=head2 ok_memcache()

Verify that memcache is running, and skip all tests if it isn't.

	my $memcache = LocalTest::ok_memcache();

=cut

sub ok_memcache
{
	# Verify if Cache::Memcached::Fast is installed.
	eval 'use Cache::Memcached::Fast';
	plan( skip_all => 'Cache::Memcached::Fast required to test cache-related features.' ) && exit
		if $@;

	my $memcache = LocalTest::get_memcache();

	# Verify that memcache is configured and running.
	plan( skip_all => 'Memcache is not running or configured on this machine, cannot test.' ) && exit
		if !defined( $memcache) || !$memcache->set( 'test_ninja_orm', 1, time() + 10 );

	return $memcache;
}


=head1 AUTHOR

Guillaume Aubert, C<< <aubertg at cpan.org> >>.


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc LocalTest


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-NinjaORM>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-NinjaORM>

=item * MetaCPAN

L<https://metacpan.org/release/DBIx-NinjaORM>

=back


=head1 COPYRIGHT & LICENSE

Copyright 2009-2014 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


t/lib/TestSubclass/NoPK.pm
package TestSubclass::NoPK;

use strict;
use warnings;

use base 'DBIx::NinjaORM';


=head1 NAME

TestSubclass::NoPK - Test subclass without a primary key.


=head1 VERSION

Version 3.0.2

=cut

our $VERSION = '3.0.2';


=head1 SYNOPSIS

	use lib 't/lib';
	use TestSubclass::NoPK;


=head1 DESCRIPTION

Test subclass without a primary key name defined, which should not allow ORM
operations.


=head1 FUNCTIONS

=head2 static_class_info()

Configure static class information.

=cut

sub static_class_info
{
	my ( $class ) = @_;

	my $info = $class->SUPER::static_class_info();

	$info->set(
		{
			table_name => 'tests',
		}
	);

	return $info;
}

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.010;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'DBIx::NinjaORM',
  'VERSION_FROM' => 'lib/DBIx/NinjaORM.pm',
  'PREREQ_PM' => {
                   'Carp' => 0,
                   'Class::Load' => 0,
                   'DBD::SQLite' => 0,
                   'DBI' => 0,
                   'DBIx::Inspector' => 0,
                   'Data::Dumper' => 0,
                   'Data::Validate::Type' => 0,
                   'Digest::SHA1' => 0,
                   'Log::Any' => 0,
                   'MIME::Base64' => 0,
                   'Storable' => 0,
                   'Test::Deep' => 0,
                   'Test::Exception' => 0,
                   'Test::FailWarnings' => 0,
                   'Test::More' => '0.94',
                   'Test::Type' => 0,
                   'Test::Warn' => 0,
                   'Try::Tiny' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t t/050-StaticClassInfo/*.t t/lib/*.t t/lib/TestSubclass/*.t t/025-Utils/*.t t/SQL/*.t t/030-Schema-Table/*.t t/010-LocalTest/*.t'
            }
)
;


lib/DBIx/NinjaORM/Utils.pm
package DBIx::NinjaORM::Utils;

use 5.010;

use strict;
use warnings;

use Carp;
use Data::Dumper ();
use Data::Validate::Type;

use base 'Exporter';

our @EXPORT_OK = qw(
	dumper
);


=head1 NAME

DBIx::NinjaORM::Utils - Utility functions for L<DBIX::NinjaORM>.


=head1 VERSION

Version 3.0.2

=cut

our $VERSION = '3.0.2';


=head1 DESCRIPTION

Collection of utility functions used by L<DBIX::NinjaORM> to perform various
ancillary tasks.


=head1 SYNOPSIS

	use DBIx::NinjaORM::Utils qw( dumper );

	my $string = dumper( $data_structure );


=head1 FUNCTIONS

=head2 dumper()

Utility to stringify data structures.

	my $string = DBIx::NinjaORM::Utils::dumper( @data );

Internally, this uses Data::Dumper::Dumper, but you can switch it to a custom
dumper with the following code:

	local $DBIx::NinjaORM::Utils::DUMPER = sub
	{
		my ( @refs ) = @_;

		# Create a stringified version.

		return $string;
	};

=cut

our $DUMPER = undef;

sub dumper
{
	my ( @data ) = @_;

	if ( defined( $DUMPER ) )
	{
		carp "The custom dumper function is not a valid code reference"
			if !Data::Validate::Type::is_coderef( $DUMPER );

		return $DUMPER->( @data );
	}
	else
	{
		return Data::Dumper::Dumper( @data );
	}
}


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc DBIx::NinjaORM::Utils


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-NinjaORM>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-NinjaORM>

=item * MetaCPAN

L<https://metacpan.org/release/DBIx-NinjaORM>

=back


=head1 AUTHOR

Guillaume Aubert, C<< <aubertg at cpan.org> >>.


=head1 COPYRIGHT & LICENSE

Copyright 2009-2014 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;



lib/DBIx/NinjaORM/Schema/Table.pm
package DBIx::NinjaORM::Schema::Table;

use strict;
use warnings;

use Carp;
use DBIx::Inspector;


=head1 NAME

DBIx::NinjaORM::Schema::Table - Store information about a table used by L<DBIx::NinjaORM>.


=head1 VERSION

Version 3.0.2

=cut

our $VERSION = '3.0.2';


=head1 DESCRIPTION

L<DBIx::NinjaORM::Schema::Table> currently uses L<DBIx::Inspector> to retrieve
various information about the tables used. This may however change in the
future, so this package encapsulates various functions to make it easier to
replace the dependencies and internals later if needed.


=head1 SYNOPSIS

	use DBIx::NinjaORM::Schema::Table;
	my $table_schema = DBIx::NinjaORM::Schema::Table->new(
		dbh  => $dbh,
		name => $name,
	);

	my $column_names = $table_schema->get_column_names();


=head1 METHODS

=head2 new()

Create a new DBIx::NinjaORM::Schema::Table object.

	my $table_schema = DBIx::NinjaORM::Schema::Table->new(
		dbh  => $dbh,
		name => $name,
	);

Arguments:

=over 4

=item * dbh (mandatory)

The database handle to use to access the table.

=item * name (mandatory)

The name of the table.

=back

=cut

sub new
{
	my ( $class, %args ) = @_;
	my $name = delete( $args{'name'} );
	my $dbh = delete( $args{'dbh'} );
	croak 'The following arguments are not valid: ' . join( ', ', keys %args )
		if scalar( keys %args ) != 0;

	# Check for the mandatory parameters.
	croak 'The argument "name" is mandatory'
		if !defined( $name );
	croak 'The argument "dbh" is mandatory'
		if !defined( $dbh );

	# Return a blessed object.
	return bless(
		{
			name => $name,
			dbh  => $dbh,
		},
		$class,
	);
}


=head2 get_name()

Return the table name.

	my $table_name = $table_schema->get_name();

=cut

sub get_name
{
	my ( $self ) = @_;

	return $self->{'name'};
}


=head2 get_dbh()

Return the database handle associated with the table.

	my $dbh = $table_schema->get_dbh();

=cut

sub get_dbh
{
	my ( $self ) = @_;

	return $self->{'dbh'};
}


=head2 get_column_names()

Return the name of the columns that exist in the underlying table.

	my $column_names = $table_schema->get_column_names();

=cut

sub get_column_names
{
	my ( $self ) = @_;

	my $columns = $self->get_columns();

	return [ map { $_->name() } @$columns ];
}


=head1 INTERNAL METHODS

Warning: the API for the internal methods may change in the future. Use or
subclass with caution.


=head2 get_inspector()

Return a cached L<DBIx::Inspector object>.

	my $inspector = $table_schema->get_inspector();

=cut

sub get_inspector
{
	my ( $self ) = @_;

	if ( !defined( $self->{'inspector'} ) )
	{
			$self->{'inspector'} = DBIx::Inspector->new( dbh => $self->get_dbh() );
			croak 'Failed to create the DBIx::Inspector object'
				if !defined( $self->{'inspector'} );
	}

	return $self->{'inspector'};
}


=head2 get_table()

Return the cached L<DBIx::Inspector::Table> object associated with the
underlying table.

	my $table = $table_schema->get_table();

=cut

sub get_table
{
	my ( $self ) = @_;

	if ( !defined( $self->{'table'} ) )
	{
		my $inspector = $self->get_inspector();
		$self->{'table'} = $inspector->table(
			$self->get_name()
		);

		croak 'Failed to retrieve the table object'
			if !defined( $self->{'table'} );
	}

	return $self->{'table'};
}

=head2 get_columns()

Return the cached arrayref of L<DBIx::Inspector::Column> objects corresponding
to the columns of the underlying table.

	my $columns = $table_schema->get_columns();

=cut

sub get_columns
{
	my ( $self ) = @_;

	if ( !defined( $self->{'columns'} ) )
	{
		my $table = $self->get_table();
		$self->{'columns'} = [ $table->columns() ];
	}

	return $self->{'columns'};
}


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc DBIx::NinjaORM::Schema::Table


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/DBIx-NinjaORM/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-NinjaORM>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-NinjaORM>

=item * MetaCPAN

L<https://metacpan.org/release/DBIx-NinjaORM>

=back


=head1 AUTHOR

Guillaume Aubert, C<< <aubertg at cpan.org> >>.


=head1 COPYRIGHT & LICENSE

Copyright 2009-2014 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


Build.PL
use 5.006;
use strict;
use warnings;

use Module::Build;


my $class = Module::Build->subclass(
	code =>
	q|
		# Override 'test' to support running extended tests in the xt/
		# directory.
		sub ACTION_test
		{
			my ( $self ) = @_;

			$self->recursive_test_files(1);

			$self->test_files( 't', 'xt' )
				if $ENV{'RELEASE_TESTING'};

			return $self->SUPER::ACTION_test();
		}

		# Force running extended tests when testing the distribution.
		sub ACTION_disttest
		{
			my ( $self ) = @_;

			local $ENV{ RELEASE_TESTING } = 1;

			return $self->SUPER::ACTION_disttest();
		}
	|,
);

my $builder = $class->new(
	module_name          => 'DBIx::NinjaORM',
	license              => 'GPL_3',
	dist_author          => q{Guillaume Aubert <aubertg@cpan.org>},
	dist_version_from    => 'lib/DBIx/NinjaORM.pm',
	build_requires       =>
	{
		'DBD::SQLite'          => 0,
		'DBI'                  => 0,
		'Test::Deep'           => 0,
		'Test::Exception'      => 0,
		'Test::FailWarnings'   => 0,
		'Test::More'           => 0.94, # subtest()
		'Test::Type'           => 0,
		'Test::Warn'           => 0,
	},
	requires             =>
	{
		'Carp'                 => 0,
		'Class::Load'          => 0,
		'DBIx::Inspector'      => 0,
		'Data::Dumper'         => 0,
		'Data::Validate::Type' => 0,
		'Digest::SHA1'         => 0,
		'Log::Any'             => 0,
		'MIME::Base64'         => 0,
		'Storable'             => 0,
		'Try::Tiny'            => 0,
		'perl'                 => 5.010,
	},
	add_to_cleanup       =>
	[
		'DBIx-NinjaORM-*',
	],
	create_makefile_pl   => 'traditional',
	configure_requires   =>
	{
		'Module::Build'        => 0,
	},
	meta_merge           =>
	{
		resources              =>
		{
			repository  => 'https://github.com/guillaumeaubert/DBIx-NinjaORM',
			homepage    => 'https://metacpan.org/release/DBIx-NinjaORM',
			bugtracker  => 'https://github.com/guillaumeaubert/DBIx-NinjaORM/issues',
		},
		no_index               =>
		{
			file        =>
			[
				't/lib/LocalTest.pm',
			],
			namespace   =>
			[
				'TestSubclass',
			]
		},
	},
	recursive_test_files => 1,
);

$builder->create_build_script();
